keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -666,7 +773,1310 @@ ways.  One is that we are dispatching requests to be served rather
+  and as large as `N` (when every queue gets its full demand).
+
+- Maintain the quantity `CL = rho_K(t) * (N-L) + Sum[over j < L]
+  rho(RI(j),t)` where `rho_K(t)` is `rho(RI(L-1),t)` when `L>0` and
+  `0` otherwise.  This is the capacity usage implied by all the fully
+  granted demands, including the implication that the queues with
+  higher demands get at least the highest fully granted demand.
+
+- The above imply that when `L < N`, `L` has the property that making
+  it one bigger would blow `CL` past the bound `C`.
+
+- Given the above, `mu_fair(t)` is:
+
+  - `C / N` if `L = 0`
+  
+  - `rho(RI(L-1),t) + (C - CL) / (N - L)` if `0 < L < N`
+
+  - `rho(RI(N-1),t)` if `L = N`.
+
+The permutation `RI` and its inverse can be implemented in such a way
+that every operation we invoke on them costs no more than `O(log N)`
+time.  For example, use a [https://github.com/wangjia184/sortedset]
+where the keys are stringified `i` and the scores are `rho(i,t)`,
+making the rank for `i` be `IR(i) + 1` and thus
+`GetByRank(j+1,false).Score()` implements `rho(RI(j),t)`.
+
+With that representation, the time complexity of solving the next
+instance of the max-min fair allocation problem is `O((1 + N_Delta) *
+log N)` where `N_Delta` is the number of queues that cross the
+boundary.  Note that while the system stays out of overload, `N_Delta`
+is zero.  It also stays zero while the system stays overloaded by the
+same set of queues.
+
+##### Serial scheduling in the virtual world
+
+Next consider the problem of scheduling in the virtual world.  This is
+a matter of computing the `B` and `E` values in the virtual world.
+This is an on-line problem, and the relevant input events that come in
+from the real world are:
+- arrival of a request into a queue,
+- completion of execution of a request in the real world (causing an
+  update to its `len`), and
+- ejecting a request that, in the real world, is waiting in a queue
+  (in the virtual world this request could be waiting or executing).
+
+Requests get dispatched in the virtual world at the same moment when
+some request completes in the virtual world.  Note that a request's
+completion in the virtual world is a function of the fictitious
+execution schedule in the that world and so generally does _not_
+correspond in time with any real world event.
+
+Note also that aside from the cases where a request starts running as
+soon as it arrives, the implementation is dealing with estimates of
+what will happen in the future.  These estimates are made under the
+assumption that nothing else will change from now to then, and later
+updated as more information comes in.  Sadly, sometimes relevant
+information comes in after it was first needed.  [TODO: summarize
+coping better.]
+
+The following subsections cover a progression of solutions for virtual
+world scheduling of one request at a time per queue.
+
+##### Simplest virtual-world serial scheduling implementation
```

Comment:
Marker to myself - this is where I got to with reading so far. Will continue over the weak.